Draft of the blog post about Core Solidity (#225)
argotorg/solcore: blog-post.md

incorporating rigorous safety guarantees inspired by modern systems programming 
languages such as Rust. Although this architectural direction represents a 
significant evolution, it keeps the domain-specific requirements of smart 
contract development, preserving the established conventions and practical 
utility of Classic Solidity where appropriate.

The language's theoretical foundation comprises several formally-specified 
features that provide enhanced type safety and expressivity. In what follows, we 
provide an overview of such features. However, we warn the reader that the current 
Core Solidity prototype uses a syntax which is similar, but not identical, to 
Classic Solidity. Since our initial concern is to develop the new language type 
system and its semantics, its surface syntax would probably change in the near 
future. All presented examples work in the current prototype.

### Generics and Type Classes

Core Solidity will introduce two new exciting abstraction mechanisms: generics and 
type classes.

Generics enable parametric polymorphism through type parameters, which allows the  
implementation of algorithms and data structures that operate uniformly across 

> over all types. As an example, we could define a polymorphic `identity` function:

Reviewer: ```suggestion
implementation of algorithms and data structures that operate uniformly across 
all types. As an example, we could define a polymorphic `identity` function:
```

``` 
forall a . function identity(x : a) -> a {
   return x;
}
```

While generic functions are interesting, most interesting operations are not defined 
at all types. Overloading allows the definition of code which can operate in distinct 
ways at different types. Type classes are the standard way of combining overloading and 
parametric polymorphism (generics) in a systematic manner. A type class definition 
declares the class name, its arguments and member functions type signature. As an 
example, let's consider the task of defining addition over different types:
```
function a . class a : Sum {
  function sum (x : a, y : a) -> a;
}
```
Implementations for member functions for different types are provided by instance 
definitions. As an example, let's consider the implementation of `Sum` for 
`word` type which, internally, uses Yul assembly to perform addition. 
```
instance word : Sum {

>   function sum (x : word, y : word) -> word {

Reviewer: Can we make the example a bit more practical by doing a sum of two arbitrary `uintX` types? Or a concatenation of two arbitrary `bytesN` types?

The current examples are ok as illustrations of the concept but we should also use them to make people a bit more excited about Core Solidity by showing how easy it is to do stuff that is really tedious and repetitive in Classic Solidity.

current highest bid and the address that made such bid, `Ended` represents that the auction has 
finished with success and it holds the highest bid and the winner address and constructor 
`Cancelled` is used when the auction has been cancelled.

Using algebraic data types, we can define function by pattern matching. As an example, 
consider function `processAuction` which 
Pattern matching provides structural decomposition of data types through case analysis. This 
ensures all possible variants are handled explicitly, eliminating partial function hazards and 
providing formal guarantees of match completeness.

```
function processAuction(state) {
    match state {
    |   NotStarted(reserve) => {
            require(msg.value >= reserve, "Bid below reserve");
            return Active(msg.value, msg.sender);
        }
    |   Active(currentBid, bidder) => {
            require(msg.value > currentBid, "Bid too low");
            transferFunds(bidder, currentBid); // refund 
            return Active(msg.value, msg.sender);
        }
    |   Cancelled => {
            revert("Auction cancelled");
        }

> ]   |   _ => {

Reviewer: Stray `]`?
```suggestion
    |   _ => {
```

            return state;
        }
    }
}
```

### High-order functions

Functions possess first-class status within the type system, enabling their use
as parameters, return values, and assignable entities. This facilitates the
implementation of higher-order functions and functional composition patterns,
enhancing expressive capability while maintaining referential transparency.
As an example of a high-order function, let's consider `map`, which applies a 
given function to all elements of an memory array.

```

> forall T U . function map (memory(array(T)) input, transform : (T) -> U) : memory(array(U)) {

Reviewer: I have trouble parsing this.

First `memory(array(T)) input`. It looks kinda as if you're using prefix typing here. Was it meant to be `input : memory(array(T))`?

Second `: memory(array(U))`. I think this is the return type of the function, but what makes me unsure is that it uses `:` rather than `->` as previous examples did. Is it a mistake or is it actually not a return type?

>     let result memory(array(U)) = new(memory(array(U)),input.length);

Reviewer: Isn't this missing `:` between the name and type?
```suggestion
    let result : memory(array(U)) = new(memory(array(U)), input.length);
```

Such patterns of data structure manipulation can be elegantly coded as 
polymorphic high-order functions, thus enabling more code reuse and less 
testing effort.

### Type inference

Core Solidity uses type inference algorithm to reduce syntactic verbosity while 
maintaining the strong static typing guarantees. In our view, static typing is essential 
for secure smart contract development. The type inference occurs during compilation and 
provides complete type safety without explicit annotations.

```
let constant_value = 42; // Inferred as word
let heterogeneous_tuple = (address(0), true, 42); // Inferred as (address, bool, word)
constant_value = heterogeneous_tuple ; // this is a type error!
```

### Core Solidity and SAIL

The combination of all these features do form a complete programming language, 
which we call SAIL (Solidity Abstract Intermediate Language). Core Solidity is 
built on top of SAIL by providing more specialized syntactic constructs that 
replicate familiar features from Classic Solidity, alongside a mechanism for 
"desugaring" them into SAIL equivalent code. But what Core Solidity features 

> will be translated into SAIL?

Reviewer: As for things that I'd actually expect to see as examples of desugaring, how about contracts and external dispatch? Or interfaces?

> will be translated into SAIL?

Reviewer: I see we have a contract example now, which is great.

What about my first comment though? You marked this as resolved without answering, so I'm not sure if you think it's not a good idea or if you simply missed it.

> will be translated into SAIL?

Author: I added the ranged loops as a feature that will be implemented in Core Solidity as a desugaring step.

- Data Locations represented as types: Data locations (e.g., storage, memory) 
are now a part of a type. This enables the creation of composite types with 
mixed locations, such as a struct that holds references to storage 
arrays and memory data, as presented in the next code piece:

```
struct Broker {
  total : memory(word) 
, accounts : storage(array(address, word))
}
```

- Static/dynamic array - literals: Classic Solidity limits array literals to static 
arrays. Core Solidity type system allows literals to adopt the correct type (static 
or dynamic) from their usage context from the start.

The current Core Solidity prototype support all previous features. In the near 
future, we plan to develop: 

- Compile time evaluation: Inspired by Zig, we plan to include compile-time code 
evaluation in Core Solidity. Details on how such feature will work are being 
discussed by Argot Collective Programming Languages research team.

- Redesigned Module System: Currently, the Core Solidity prototype does not have 
a proper module system. In our tests, we implemented a import directive which 
behaves like C's `#include`. The design of a proper module system for Core Solidity 

> are being discussed by Argot Collective Programming Languages research team.

Reviewer: From our discussions it looks like Core Solidity can live with a module system very similar to what we have in Classic Solidity and there will be no big redesign. The change seems more about solcore prototype specifically, so I'd not mention it here as new a feature to be added. It would fit more in some section about limitations of the prototype (if it has to be mentioned at all).

- Macros: One of the main design goals of Core Solidity is to reduce the burden 
of extending and maintaining the Solidity Compiler. Currently, the Core Solidity 
prototype has a internal implementation of all desugaring steps into SAIL into the 
compiler itself, which conflicts with our design goal of having the most of language 
features being part of the standard library itself. One way to achieve such goal 
is by having a powerful **macro** system, like Lean 4 or modern Lisp dialects such 
as Racket. Having macros, most of the desugaring steps would be implemented by 
macros, allowing such steps to be part of libraries, giving users the power 
to replace standard library implementations to code that better suit their needs.
The exact details of how such macro system will be discussed in the near future 

> by Argot Collective Programming Languages research team.

Reviewer: Wasn't the macro system something to be done only post 1.0? Or will we actually be designing it in the near future?
